src/utils/allocators.zig

```diff
@@ -0,0 +1,116 @@
+const std = @import("std");
+
+pub const RecycleFBA = struct {
+    // this allocates the underlying memory + dynamic expansions
+    // (only used on init/deinit + arraylist expansion)
+    backing_allocator: std.mem.Allocator,
+    // this does the data allocations (data is returned from alloc)
+    alloc_allocator: std.heap.FixedBufferAllocator,
+    // recycling depot
+    records: std.ArrayList(Record),
+
+    // for thread safety
+    mux: std.Thread.Mutex = .{},
+
+    const Record = struct { is_free: bool, buf: []u8 };
+    const Self = @This();
+
+    pub fn init(backing_allocator: std.mem.Allocator, n_bytes: u64) !Self {
+        const buf = try backing_allocator.alloc(u8, n_bytes);
+        const alloc_allocator = std.heap.FixedBufferAllocator.init(buf);
+        const records = std.ArrayList(Record).init(backing_allocator);
+
+        return .{
+            .backing_allocator = backing_allocator,
+            .alloc_allocator = alloc_allocator,
+            .records = records,
+        };
+    }
+
+    pub fn deinit(self: *Self) void {
+        self.backing_allocator.free(self.alloc_allocator.buffer);
+        self.records.deinit();
+    }
+
+    pub fn allocator(self: *Self) std.mem.Allocator {
+        return std.mem.Allocator{
+            .ptr = self,
+            .vtable = &.{
+                .alloc = alloc,
+                .resize = resize,
+                .free = free,
+            },
+        };
+    }
+
+    /// creates a new file with size aligned to page_size and returns a pointer to it
+    pub fn alloc(ctx: *anyopaque, n: usize, log2_align: u8, return_address: usize) ?[*]u8 {
+        _ = log2_align;
+        _ = return_address;
+        const self: *Self = @ptrCast(@alignCast(ctx));
+
+        if (n > self.alloc_allocator.buffer.len) {
+            @panic("RecycleFBA.alloc: requested size too large, make the buffer larger");
```

Comment:
a common pattern with the RecycleFBA is to try to alloc, if it fails, wait for some amount of time and then try again (hoping something was freed): this is how we use it - if you try to alloc a slice greater than whats available using this approach, you will inf loop

Reply:
It's weird though for this code to panic based on some assumption about how some far away code is going to use an allocator. In the context that you're writing code that might become an infinite loop, you already have the special knowledge about the allocator. That's why the loop is there. So I think it makes sense to check for this case in the loop.

Reply:
the problem i see is assuming special knowledge about the allocator is dangerous and easy to miss/forget about, being such a killer-case (will hard stuck and be difficult to debug) might as well be panic and give a clear message to the dev

Reply:
> assuming special knowledge about the allocator is dangerous

Yeah, that's my point.

There is a loop that would normally be expected to have a high risk of running infinitely. To mitigate that risk, the loop relies on an assumption that the unique properties of this particular allocator will break out of the loop.

At that point, the loop no longer depends on just any old `Allocator`. It depends on this specific allocator. So it's only thinly disguised behind the `Allocator` interface. Plus, the allocator itself is written in a way that is specifically designed to deal with this particular looping pattern. Each piece of code is tightly coupled to specific details about how the other piece of code is written. The abstraction is broken in both directions and it might be better to implement it without the `Allocator` interface at all, so the special behavior that's already required is made more obvious.

I think it would be better for the loop itself to be written to be safe, rather than having an inherently dangerous loop that can only be saved by an allocator that's written with this particular loop in mind.